src/gdb_packet.c

```diff
@@ -31,119 +31,220 @@
 
 #include <stdarg.h>
 
+typedef enum packet_state {
+	PACKET_IDLE,
+	PACKET_GDB_CAPTURE,
+	PACKET_GDB_ESCAPE,
+	PACKET_GDB_CHECKSUM_UPPER,
+	PACKET_GDB_CHECKSUM_LOWER,
+	PACKET_GDB_COMPLETE,
+} packet_state_e;
+
+static bool noackmode = false;
+
+void gdb_set_noackmode(bool enable)
+{
+	/* 
+	 * If we were asked to disable no ack mode, and it was previously enabled,
+	 * it might mean we got a paket we determined to be the first of a new
+	 * GDB session, and as such it was not acknoledged (before GDB enabled no ack mode),
+	 * better late than never.
+	 * 
+	 * If we were asked after the connection was terminated, sending the ack have no effect.
+	 */
+	if (enable == false && noackmode == true) {
+		gdb_if_putchar(GDB_PACKET_ACK, 1U);
+	}
+
+	DEBUG_GDB("%s no ack mode\n", enable ? "Enable" : "Disable");
+
+	noackmode = enable;
+}
+
+packet_state_e consume_remote_packet(char *const packet, const size_t size)
+{
+	/* We got what looks like probably a remote control packet */
+
+#if PC_HOSTED == 0
+	size_t offset = 0;
+
+	while (true) {
+		/* Consume bytes until we either have a complete remote control packet or have to leave this mode */
+		const char rx_char = gdb_if_getchar();
+
+		switch (rx_char) {
+		case '\x04':
+			/* EOT (end of transmission) - connection was closed */
+			return PACKET_IDLE;
+
+		case REMOTE_SOM:
+			/* Oh dear, restart remote packet capture */
+			offset = 0;
+			break;
+
+		case REMOTE_EOM:
+			/* Complete packet for processing */
+
+			/* Null terminate packet */
+			packet[offset] = '\0';
+
```

Comment:
Think it's ok to drop this extra new line too as calling into remote_packet_process() feels grouped to finishing setting up the buffer for the call